### DISCUSSION TITLE
Synchronizing reported frame rate for 2 cameras?

### DISCUSSION THREAD

**Discussion by meganalo** [bonsai-rx/bonsai#1661] [Q&A]
Hi,

We have 2 USB 3.0 cameras (Imaging Source DMK 37BUR0234 + their drivers) plugged into SS ports, computer has i9 core, 2 GHz CPU, 16 GB RAM, 64 bit. We are using the VideoCaptureDevice node where we can change the resolution but we are not reliably able to acquire both high resolution and high frame rate when both cameras are recording at the same time. (Note: The frame rate can be adjusted through their recommended software 'IC capture' but we seem to only have the option to decrease resolution in Bonsai). We save separate video and csv files for each video source. When we record at 100 fps and 1920x1200 the number of frames recorded in the csv is unreliable and differs between cameras (concerningly one of the cameras appears to “skip” earlier frames as shown in screenshot). The idea will be to merge these video streams so want to ensure the frames are synchronized/ videos are the same length/ etc.  Periodically, the videos will be of vastly different sizes despite frame rate being the same so we are wondering if the computer is unable to support recording 1920x1200 at 100 fps for 2 cameras.

Any ideas what the issue might be?

<img width="700" alt="Screen Shot 2024-01-19 at 3 39 54 PM" src="https://github.com/bonsai-rx/bonsai/assets/128488461/59ef503a-6c51-414b-96a8-b75b08377350">
<img width="700" alt="Screen Shot 2024-01-19 at 3 40 22 PM" src="https://github.com/bonsai-rx/bonsai/assets/128488461/bb668f2f-88bc-4059-840c-63e93251462a">
<img width="700" alt="Screen Shot 2024-01-19 at 3 44 35 PM" src="https://github.com/bonsai-rx/bonsai/assets/128488461/98676b52-4cdf-47c3-9fa0-f5c54084249a">
<img width="700" alt="Screen Shot 2024-01-19 at 3 44 57 PM" src="https://github.com/bonsai-rx/bonsai/assets/128488461/aa15a15f-ebc2-485c-9686-a26f4390755a">

**Comment by bruno-f-cruz** [comment]
In general if you want to align cameras you should probably consider a hardware level solution. Regardless, from your csv it is not clear to me if the cameras are progressively drifting to each other or just a hiccup at the start. Is the drift significantly different among the two cameras? I would expect some variance after a few seconds (eg if one of them is going at 99 and another 101fps) but not huge.

**Reply by meganalo** [reply]
Good to know about aligning the cameras, thanks. There is the hiccup at the start where cam0 jumps from 00:00:0 to 00:00:2 , then some small drift of a tenth of second throughout. At the end, cam1 cuts at 00:10:2 and cam0 runs until 00:10:5. So I suppose the variance is not very large. I wonder if a hardware solution is the answer then. 

[TimestampsOfCameraFrame0.csv](https://github.com/bonsai-rx/bonsai/files/14012221/TimestampsOfCameraFrame0.csv)
[TimestampsOfCameraFrame1.csv](https://github.com/bonsai-rx/bonsai/files/14012222/TimestampsOfCameraFrame1.csv)

**Comment by bruno-f-cruz** [comment]
If you want to get rid of the hiccup (highly likely to happen at the time of subscription to the camera observable since bonsai needs to initialize the connection to it) you can use the following pattern:


![image](https://github.com/bonsai-rx/bonsai/assets/7049351/fae11f98-a69f-4119-9ab0-ec50d55ff2f3)

<details>
    <summary>Workflow (Copy+Paste me)</summary>
    
```xml

<?xml version="1.0" encoding="utf-8"?>
<WorkflowBuilder Version="2.8.1"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                 xmlns:cv="clr-namespace:Bonsai.Vision;assembly=Bonsai.Vision"
                 xmlns:rx="clr-namespace:Bonsai.Reactive;assembly=Bonsai.Core"
                 xmlns="https://bonsai-rx.org/2018/workflow">
  <Workflow>
    <Nodes>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="cv:CameraCapture">
          <cv:Index>0</cv:Index>
          <cv:CaptureProperties />
        </Combinator>
      </Expression>
      <Expression xsi:type="rx:PublishSubject">
        <Name>Camera1</Name>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="cv:CameraCapture">
          <cv:Index>0</cv:Index>
          <cv:CaptureProperties />
        </Combinator>
      </Expression>
      <Expression xsi:type="rx:PublishSubject">
        <Name>Camera2</Name>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Timer">
          <rx:DueTime>PT5S</rx:DueTime>
          <rx:Period>PT0S</rx:Period>
        </Combinator>
      </Expression>
      <Expression xsi:type="rx:PublishSubject">
        <Name>StartTrigger</Name>
      </Expression>
      <Expression xsi:type="SubscribeSubject">
        <Name>Camera1</Name>
      </Expression>
      <Expression xsi:type="SubscribeSubject">
        <Name>StartTrigger</Name>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:SubscribeWhen" />
      </Expression>
      <Expression xsi:type="GroupWorkflow">
        <Name>DoStuff()</Name>
        <Workflow>
          <Nodes>
            <Expression xsi:type="WorkflowInput">
              <Name>Source1</Name>
            </Expression>
            <Expression xsi:type="WorkflowOutput" />
          </Nodes>
          <Edges>
            <Edge From="0" To="1" Label="Source1" />
          </Edges>
        </Workflow>
      </Expression>
      <Expression xsi:type="SubscribeSubject">
        <Name>Camera2</Name>
      </Expression>
      <Expression xsi:type="SubscribeSubject">
        <Name>StartTrigger</Name>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:SubscribeWhen" />
      </Expression>
      <Expression xsi:type="GroupWorkflow">
        <Name>DoStuff()</Name>
        <Workflow>
          <Nodes>
            <Expression xsi:type="WorkflowInput">
              <Name>Source1</Name>
            </Expression>
            <Expression xsi:type="WorkflowOutput" />
          </Nodes>
          <Edges>
            <Edge From="0" To="1" Label="Source1" />
          </Edges>
        </Workflow>
      </Expression>
    </Nodes>
    <Edges>
      <Edge From="0" To="1" Label="Source1" />
      <Edge From="2" To="3" Label="Source1" />
      <Edge From="4" To="5" Label="Source1" />
      <Edge From="6" To="8" Label="Source1" />
      <Edge From="7" To="8" Label="Source2" />
      <Edge From="8" To="9" Label="Source1" />
      <Edge From="10" To="12" Label="Source1" />
      <Edge From="11" To="12" Label="Source2" />
      <Edge From="12" To="13" Label="Source1" />
    </Edges>
  </Workflow>
</WorkflowBuilder>

```

</details>
